external-crates/move/documentation/book/src/control-flow/pattern-matching.md

```diff
@@ -0,0 +1,618 @@
+        Pair(true, false) => 1,
+        Pair(false, false) => 1,
+        // ERROR: not exhaustive as the value `Pair(false, true)` is not matched.
+    }
+}
+```
+
+These examples can then be made exhaustive by adding a wildcard pattern to the end of the match arm,
+or by fully matching on the remaining values:
+
+```move
+fun f(x: MyEnum): u8 {
+    match (x) {
+        MyEnum::Variant(1, true) => 1,
+        MyEnum::Variant(_, _) => 1,
+        MyEnum::OtherVariant(_, 3) => 2,
+        // Now exhaustive since this will match all values of MyEnum::OtherVariant
+        MyEnum::OtherVariant(..) => 2,
+
+    }
+}
+
+fun match_pair_bool(x: Pair<bool>): u8 {
+    match (x) {
+        Pair(true, true) => 1,
+        Pair(true, false) => 1,
+        Pair(false, false) => 1,
+        // Now exhaustive since this will match all values of Pair<bool>
+        Pair(false, true) => 1,
+    }
+}
+```
+
+## Guards
+
+As mentioned above you can add a guard to a match arm by adding an `if` clause after the pattern.
+This guard will run _after_ the pattern has been matched but _before_ the expression on the
+right-hand-side of the arrow is evaluated. If the guard expression evaluates to `true` then the
+expression on the right-hand side of the arrow will be evaluated, if it evaluates to `false` then it
+will be considered a "failed match" and the next match arm in the `match` expression will be
+checked.
+
+```move
+fun match_with_guard(x: u64): u64 {
+    match (x) {
+        1 if (false) => 1,
+        1 => 2,
+        _ => 3,
+    }
+}
+
+match_with_guard(1); // returns 2
+match_with_guard(0); // returns 3
+```
+
+Guard expressions can reference variables bound in the pattern during evaluation. However, note that
+_variables are only available as immutable reference in guards_ regardless of the pattern being
+matched -- even if there are mutability specifiers on the variable or if the pattern is being
+matched by value.
+
+```move
+fun incr(x: &mut u64) {
+    *x = *x + 1;
+}
+
+fun match_with_guard_incr(x: u64): u64 {
+    match (x) {
+        x if ({ incr(&mut x); x == 1 }) => 1,
+        // ERROR:    ^^^ invalid borrow of immutable value
+        _ => 2,
+    }
+}
+
+fun match_with_guard_incr2(x: &mut u64): u64 {
+    match (x) {
+        x if ({ incr(&mut x); x == 1 }) => 1,
+        // ERROR:    ^^^ invalid borrow of immutable value
+        _ => 2,
+    }
+}
+```
+
+Additionally, it is important to note any match arms that have guard expressions will not be
+considered either for exhaustivity purposes since the compiler has no way of evaluating the guard
+expression statically.
+
+## Limitations on Specific Patterns
+
+There are some restrictions on when the `..` pattern, and `mut` pattern modifier can be used in a
+pattern.
+
+### Mutability Usage
+
+A `mut` modifier can be placed on a variable pattern to specify that the _variable_ is to be mutated
+in the right-hand side expression of the match arm. Note that since the `mut` modifier only
```

Comment:
```suggestion
in the right-hand expression of the match arm. Note that since the `mut` modifier only
```
Dang it, now I'm committed. Had I known it was going to be a thing, I would have just removed the one dash from the original comment. Here, the dash helps but I don't think the "side" is needed.